File: apps/backend/src/auth/auth.service.ts
import {
  ConflictException,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { JwtService } from '@nestjs/jwt';
import * as bcrypt from 'bcrypt';
import { Inject } from '@nestjs/common';
import Redis from 'ioredis';
import { PrismaService } from '../prisma/prisma.service';
import { REDIS_CLIENT } from '../redis/redis.module';
import { RegisterDto } from './dto/register.dto';
import { LoginDto } from './dto/login.dto';

@Injectable()
export class AuthService {
  constructor(
    private prisma: PrismaService,
    private jwt: JwtService,
    private config: ConfigService,
    @Inject(REDIS_CLIENT) private redis: Redis,
  ) {}

  async register(dto: RegisterDto) {
    const exists = await this.prisma.user.findUnique({ where: { email: dto.email } });
    if (exists) throw new ConflictException('EMAIL_TAKEN');

    const passwordHash = await bcrypt.hash(dto.password, 10);

    const user = await this.prisma.$transaction(async (tx) => {
      const u = await tx.user.create({
        data: { email: dto.email, passwordHash, role: dto.role, gender: dto.gender },
      });
      await tx.wallet.create({ data: { userId: u.id, balance: 0 } });
      return u;
    });

    const tokens = await this.signTokens(user.id, user.role);
    await this.saveRefreshToken(user.id, tokens.refreshToken);
    return { ...tokens, userId: user.id };
  }

  async login(dto: LoginDto) {
    const user = await this.prisma.user.findUnique({ where: { email: dto.email } });
    if (!user) throw new UnauthorizedException('INVALID_CREDENTIALS');

    const valid = await bcrypt.compare(dto.password, user.passwordHash);
    if (!valid) throw new UnauthorizedException('INVALID_CREDENTIALS');

    const tokens = await this.signTokens(user.id, user.role);
    await this.saveRefreshToken(user.id, tokens.refreshToken);
    return { ...tokens, userId: user.id, role: user.role };
  }

  async refresh(userId: string, refreshToken: string) {
    const stored = await this.redis.get(`refresh:${userId}`);
    if (!stored) throw new UnauthorizedException('SESSION_EXPIRED');

    const valid = await bcrypt.compare(refreshToken, stored);
    if (!valid) throw new UnauthorizedException('INVALID_REFRESH_TOKEN');

    const user = await this.prisma.user.findUniqueOrThrow({ where: { id: userId } });
    const tokens = await this.signTokens(user.id, user.role);
    await this.saveRefreshToken(user.id, tokens.refreshToken);
    return tokens;
  }

  async logout(userId: string) {
    await this.redis.del(`refresh:${userId}`);
  }

  private async signTokens(userId: string, role: string) {
    const [accessToken, refreshToken] = await Promise.all([
      this.jwt.signAsync(
        { sub: userId, role },
        { secret: this.config.get('JWT_SECRET'), expiresIn: '15m' },
      ),
      this.jwt.signAsync(
        { sub: userId, role },
        { secret: this.config.get('JWT_REFRESH_SECRET'), expiresIn: '7d' },
      ),
    ]);
    return { accessToken, refreshToken };
  }

  private async saveRefreshToken(userId: string, refreshToken: string) {
    const hash = await bcrypt.hash(refreshToken, 10);
    await this.redis.set(`refresh:${userId}`, hash, 'EX', 604800);
  }
}

File: apps/backend-dotnet/PawPals.Api/Services/AuthService.cs
using System.IdentityModel.Tokens.Jwt;
using System.Security.Claims;
using System.Text;
using Microsoft.EntityFrameworkCore;
using Microsoft.IdentityModel.Tokens;
using StackExchange.Redis;
using PawPals.Api.Data;
using PawPals.Api.DTOs;
using PawPals.Api.Models;

namespace PawPals.Api.Services;

public class AuthService(AppDbContext db, IConnectionMultiplexer redis, IConfiguration config)
{
    private readonly IDatabase _redis = redis.GetDatabase();

    public async Task<TokenResponse> Register(RegisterDto dto)
    {
        if (await db.Users.AnyAsync(u => u.Email == dto.Email))
            throw new InvalidOperationException("EMAIL_TAKEN");

        var user = new User
        {
            Email = dto.Email,
            PasswordHash = BCrypt.Net.BCrypt.HashPassword(dto.Password),
            Role = dto.Role,
            Gender = dto.Gender,
        };

        await using var tx = await db.Database.BeginTransactionAsync();
        db.Users.Add(user);
        db.Wallets.Add(new Wallet { UserId = user.Id });
        await db.SaveChangesAsync();
        await tx.CommitAsync();

        var tokens = SignTokens(user.Id, user.Role.ToString());
        await SaveRefreshToken(user.Id, tokens.RefreshToken);
        return tokens with { UserId = user.Id };
    }

    public async Task<TokenResponse> Login(LoginDto dto)
    {
        var user = await db.Users.FirstOrDefaultAsync(u => u.Email == dto.Email)
            ?? throw new UnauthorizedAccessException("INVALID_CREDENTIALS");

        if (!BCrypt.Net.BCrypt.Verify(dto.Password, user.PasswordHash))
            throw new UnauthorizedAccessException("INVALID_CREDENTIALS");

        var tokens = SignTokens(user.Id, user.Role.ToString());
        await SaveRefreshToken(user.Id, tokens.RefreshToken);
        return tokens with { UserId = user.Id, Role = user.Role.ToString() };
    }

    public async Task<TokenResponse> Refresh(string userId, string refreshToken)
    {
        var stored = await _redis.StringGetAsync($"refresh:{userId}");
        if (!stored.HasValue) throw new UnauthorizedAccessException("SESSION_EXPIRED");
        if (!BCrypt.Net.BCrypt.Verify(refreshToken, stored!)) throw new UnauthorizedAccessException("INVALID_REFRESH_TOKEN");

        var user = await db.Users.FindAsync(userId) ?? throw new UnauthorizedAccessException();
        var tokens = SignTokens(user.Id, user.Role.ToString());
        await SaveRefreshToken(user.Id, tokens.RefreshToken);
        return tokens;
    }

    public async Task Logout(string userId) =>
        await _redis.KeyDeleteAsync($"refresh:{userId}");

    private TokenResponse SignTokens(string userId, string role)
    {
        var access = CreateToken(userId, role, config["JWT_SECRET"]!, TimeSpan.FromMinutes(15));
        var refresh = CreateToken(userId, role, config["JWT_REFRESH_SECRET"]!, TimeSpan.FromDays(7));
        return new TokenResponse(access, refresh, userId, role);
    }

    private static string CreateToken(string userId, string role, string secret, TimeSpan expires)
    {
        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(secret));
        var creds = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);
        var token = new JwtSecurityToken(
            claims: [new Claim("sub", userId), new Claim("role", role)],
            expires: DateTime.UtcNow.Add(expires),
            signingCredentials: creds);
        return new JwtSecurityTokenHandler().WriteToken(token);
    }

    private async Task SaveRefreshToken(string userId, string token)
    {
        var hash = BCrypt.Net.BCrypt.HashPassword(token);
        await _redis.StringSetAsync($"refresh:{userId}", hash, TimeSpan.FromDays(7));
    }
}
